cmd/ocsp-updater/main.go

```diff
@@ -107,96 +243,162 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
-
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
+func (updater *OCSPUpdater) missingSCTLoop() {
+	c := time.Tick(updater.missingSCTWindow)
+	for now := range c {
+		updater.rpcMu.RLock()
+
+		since := now.Add(-updater.oldestIssuedSCT)
+		serials, err := updater.getSerialsIssuedSince(since, updater.missingSCTBatch)
+		if err != nil && err != errNoNewCertificates {
+			updater.log.AuditErr(err)
+			updater.rpcMu.RUnlock()
+			continue
+		}
+
+		for _, serial := range serials {
+			// TODO(rolandshoemaker): For now this will do, in the future we should
+			// really have another method that allows us to specify which log/s should
+			// be submitted to so we don't double submit to logs we already have receipts
+			// for
+			if count, err := updater.getNumberOfReceipts(serial); err == nil && count != updater.numLogs {
+				var certDER []byte
+				updater.dbMap.SelectOne(
+					&certDER,
+					`SELECT der FROM certificates
+					 WHERE serial = :serial`,
+					map[string]interface{}{"serial": serial},
+				)
+				if err != nil {
+					updater.log.AuditErr(err)
+					continue
+				}
+
+				err = updater.pub.SubmitToCT(certDER)
+				if err != nil {
+					updater.log.AuditErr(err)
+					continue
+				}
+			} else if err != nil {
+				updater.log.AuditErr(err)
+				continue
 			}
 		}
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
+		updater.stats.TimingDuration("OCSP.TickDuration.MissingSCT", time.Since(now), 1.0)
+		updater.rpcMu.RUnlock()
 	}
-
-	return err
 }
 
-func main() {
-	app := cmd.NewAppShell("ocsp-updater", "Generates and updates OCSP responses")
+func (updater *OCSPUpdater) oldOCSPLoop() {
+	c := time.Tick(updater.oldOCSPWindow)
+	for now := range c {
+		updater.rpcMu.RLock()
+
+		certs, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry))
+		if err != nil && err != errNoStaleResponses {
```

Comment:
why does errNoStaleResponses need to be an error condition? It doesn't seem like an error to find nothing that needs to be done. In fact, if we're lucky, that will usually be true. We can just check `len(certs)` if we want to do special things for the zero case.